### ISSUE TITLE
[no-static-element-interactions] Suppress warning if specific role is specified?

### ISSUE THREAD

**Issue by tleunen** [evcohen/eslint-plugin-jsx-a11y#91]
Should the rule be auto-suppressed when there's a role attribute on the element?

``` jsx
<div role="button" onClick={sth}>Click me</div>
```

**Comment by ljharb** [comment]
If it's a button, why isn't it a button?

**Comment by tleunen** [comment]
Because of a CSS framework that style all buttons but for this specific case you don't want it to be styled as a "true" button? ^^

Note I agree if it has a click function, it should be a link or a button, but sometimes it's not possible and I thought the role was meant for that.

**Comment by lencioni** [comment]
Role is important for that, but if you want this element to be accessible and usable, you also need to give it a tabIndex so it can be focused, and provide a keyboard handler to allow space or enter to activate it.

**Comment by tleunen** [comment]
Is `<a onClick={sth}>` valid or does it require something more?

**Comment by lencioni** [comment]
That needs an `href` that points to the relevant content (e.g. not `#`).

**Comment by tleunen** [comment]
So the only valid element that can have `onClick` is the `<button>`?

**Comment by lencioni** [comment]
Also interactive form elements like input, and links.

**Comment by lencioni** [comment]
Oh, also `<area>`

**Comment by mastertinner** [comment]
I'm experiencing a similar problem like @tleunen. I currently have a couple of `<div>`s with the `role="button"` and `onClick={sth}`. These are mainly chevrons to open collapsible areas which I don't want to be styled like a button at all and which don't point to a specific URL. So I can't apply either `<button>` (messes up all styles and has different style presets depending on the browser, which I would have to explicitly overwrite for each of these interactive elements) nor `<a>` (since my elements don't point to a new URL but only open a collapsible area).

Do you guys have any suggestions for these cases? According to http://uxmovement.com/buttons/when-to-use-a-button-or-link/ they should be links from a UX perspective but I don't see how I would add a `href` pointing to something in my case.

**Comment by sir-marc** [comment]
More like a best practice question:
If I have a Modal component with a Modalcontainer Component. 
Both are divs, how do I handle an onClick event on the ModalContainer Component? I want then to close the Modal so I need an onClick handler on the div element? Do i get something terribly wrong or is this case acceptable?

**Comment by shulcsm** [comment]
What about drawable canvas?

**Comment by seansean11** [comment]
What are most people doing to remedy this?  Does this mean that we should zero out all of the CSS styles on something like the button element across browsers so that we can use it for click events.  There are so many instances in modern dev where you need to use click events on "non-interactive" elements and replacing with the button element (or any of the other "interactive elements") carries undesired styles with it by default.  I understand the need for accessibility, but can't help but feel that these "interactive" elements (and their default styles) were not originally intended to be used in the way that we are using them.

**Comment by evcohen** [comment]
There are ways to make non-interactive elements accessible (as mentioned [above](https://github.com/evcohen/eslint-plugin-jsx-a11y/issues/91#issuecomment-246408640) by @lencioni). this rule is a fairly strict rule if you want to ensure a11y across all devices without having to add these special properties to the non-interactive element. you can turn off this rule if you absolutely need to for styling purposes, but usually there is a way to style an accessible, semantic element the way you want without having to make it a `<div>`

**Comment by vdh** [comment]
@evcohen So what kind of interactive element would you recommend for something like a modal backdrop with a click-to-close `onClick`?

**Comment by seansean11** [comment]
@vdh it seems like he is suggesting using a non-interactive element and manually adding tabIndex as well as the keyboard handlers.  IMO, this sounds quite cumbersome.  @evcohen , I wasn't inferring that I would turn semantic elements into `<div>`'s.  However, what I'd like to avoid is turning perfectly semantic elements into `<button>`'s or other "interactive" elements #105, which may be less semantic for the situation.  @vdh had a great example of this.

Personally, my gripe here has nothing to do with the plugin (I'm glad it brought this issue to my attention).  I'm just trying to see if there is a better way to deal with these click events because adding custom tabindex/keyboard events or turning semantic elements into `<button>`'s both feel awkward to me.

**Comment by ljharb** [comment]
If it was "perfectly semantic", it wouldn't be a non-natively-clickable element with click behavior. Modals are not "perfectly semantic" in any scenario - so for modals, you may indeed have to use comments to override the rule.

**Comment by evcohen** [comment]
agreed @ljharb. @vdh also, there may be better ways to design escaping a modal, such as adding an accessible close button and allowing the `esc` button to close out the modal, i believe.

@seansean11 this is true- it is cumbersome, which is another reason why it's simpler and safer to use semantic, interactive elements to fit your use cases. one of the great things about React, though, is that you can create an `Accessible` component wrapper that automatically applies these props to your non-interactive `x` element (`x` being `<div>` for example). This way, you only don't have to remember to apply these props every time in the wild.

**Comment by bensleveritt** [comment]
For a good run down on how to maintain accessibility on non-semantic elements: [Mozilla have some notes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role).

**Comment by bensleveritt** [comment]
@mastertinner From the link I just posted, I think a relatively elegant solution (besides styling a button), would be to add `role="button" aria-pressed="true" tabIndex="0"`, or so.

> Additionally, this role can be used in combination with the aria-pressed attribute  to create toggle buttons.

**Comment by evcohen** [comment]
Thanks @bensleveritt! I will update the docs with this & some other highlights from this discussion.

**Comment by rubencodes** [comment]
I had similar questions to those above, and wanted to add: list-group-items within a `<ul>` are often used for menus and dropdowns (see: Bootstrap), as well as for things like contact lists (see: Messenger.com/Hangouts/etc.). Would be nice if this rule provided a way to either define a list of elements to ignore, or *even better* could be silenced with good accessibility practices like roles and tabIndexes as you all are describing above.

**Comment by Statfine** [comment]
"jsx-a11y/no-static-element-interactions": 0,

**Comment by sdwvit** [comment]
+1 should allow either use interactive elements inside other interactive elements (like backdrop in modal), or suppress warning if role is specified and use divs with onclick cb

**Comment by jillyan** [comment]
<tr
 id={id}
className={classes}
			role="button"
tabIndex="0"
onKeyDown={ () => {}}
onClick={() => onClick(item)}
onDoubleClick={() => onDoubleClick(item)}>
I added the role, tabIndex, onKeyDown, but still get the error.
that's why?

**Comment by bensleveritt** [comment]
@jillyan I don't think this change has been implemented yet. We're just at the recommendation stage.

**Comment by frassinier** [comment]
@jillyan Please consider this [implementation](http://codepen.io/frassinier/full/egvKLL/) instead. IMHO `tr` should not be clickable.

**Comment by jessebeach** [comment]
> More like a best practice question:
If I have a Modal component with a Modalcontainer Component.
Both are divs, how do I handle an onClick event on the ModalContainer Component? I want then to close the Modal so I need an onClick handler on the div element? Do i get something terribly wrong or is this case acceptable?

Put `role="presentation"` on the click catcher screen. That'll satisfy the eslint rule.

I'm working on a few updates to this rule and a couple others that tease out the nuances of when they should trigger given the inherent semantics of native HTML elements and semantic-overriding role attributes. Hopefully folks find them acceptable; stay tuned.

**Comment by jessebeach** [comment]
### Explanation

There are three semantic states that an HTML element can hold: interactive, non-interactive, static.

**Interactives** can host behavior handling.
**Non-interactives** are containers. They do not host behavior handling.
**Statics** have no semantic value. I mean this in the sense that they have no documented mapping from HTML to an Accessibility API via ARIA. For example, the `data`, `footer` and `caption` elements have no ARIA mappings. In terms of ARIA, they might as well be a `div` for all that we can know about their behavior in an assistive tech agent.

Here is a list of HTML elements grouped into one of these three categories: https://gist.github.com/jessebeach/c0e24b409418f2ffdec6d60c36553f6e

These categorizations are what drive the rules I am suggesting. 

### The plan

Here's my plan. Say an author starts with `<div onClick={} />`.

This violates the new `interactive-has-role`. I'm updating `onClick-has-role` to become `interactive-has-role`. Had this been `<button onClick={} />`, the rule would not have triggered since the inherent role of `<button>` is `button`.  We move into two scenarios from here:

#### Scenario 1: The author adds an interactive role

An interactive role is `button`, `link`, `treeitem`, etc

```
<div onClick={} role="button" />
```

This sets them down the path of making this a fully operable element: adding a `tabIndex`, a label, and support for keyboard or click interactions (whichever is missing). None of these further rules should trigger until we know what the semantics of this element are intended to be.

Had the author started with `<li onClick={} />`, this would violate `no-noninteractive-element-handlers` (discussed below), rather than `no-static-element-handlers`.

#### Scenario 2: The author adds a non-interactive role

A non-interactive role is `banner`, `list`, `dialog`, etc. Non-interactive roles designate *containers*  that contain operable elements. They should not be host-sites for handling user interaction.

```
<div onClick={} role="list" />
```

This violates `no-noninteractive-element-handlers` (new rule under development). The explanation will recommend that the author move the handlers to an element inside the container that has an appropriate, interactive role.

### The goal

We want to hold the hand of authors as they add behaviors to their HTML -- push them to use "semantic" HTML, but also offer guidance when this isn't the case. First, define the semantics and choose a role. From there, we can guide them to either building a complete interactive component or prevent them from creating an operable container with incompatible semantics. 

This ESLint plugin should surface the underlayer of semantics that ARIA provides and reveal its mysteries in actionable chunks. Otherwise we're going to get lots of `<li role="button" onClick={} onKeyPress={} tabIndex={0} />` nonsense :)

**Comment by jessebeach** [comment]
You can see the work-in-progress here: https://github.com/evcohen/eslint-plugin-jsx-a11y/compare/master...jessebeach:no-handlers-on-non-interactives

**Comment by interactivellama** [comment]
I completely get that `onClick` should be on non-interactive elements. How should one catch bubbling up keyboard events though? Seems like `<div onKeyDown />` should be allowed for custom keyboard shortcuts.

**Comment by jessebeach** [comment]
> How should one catch bubbling up keyboard events though? Seems like <div onKeyDown /> should be allowed for custom keyboard shortcuts

I agree, on* handlers on static elements like div must be allowed. My thought here is to suggest putting `role="presentation"` ( in ARIA 1.1 it would be `role="none"`) on the div with the on* handler. This indicates to the linter that this element has no impact on accessibility and should not be included in the AX Tree. What do you think?

**Comment by interactivellama** [comment]
@jessebeach I'm not an a11y expert--although my manager is asking me to be one. 😄  It seems like a good route. If I'm going to have to add role="presentation", I might as well just disable the eslint on the tag though. Also, changing markup based on where I put my event handlers vs someone else is tricky. I work on the React version of the [Lightning Design System](https://lightningdesignsystem.com/), so we've got a lot different tech stacks that implement the same markup and i'm in the process of Jest snapshot testing the markup, to keep up with markup changes.

**Comment by jessebeach** [comment]
@interactivellama oh! Do you work with @stefsullrew? We hang when she's in RWC =D

I'll run the rules I'm developing in https://github.com/evcohen/eslint-plugin-jsx-a11y/compare/master...jessebeach:no-handlers-on-non-interactives over the components in https://github.com/salesforce-ux/design-system/tree/master/ui and make sure nothing wonky and irreconcilable surfaces.

**Comment by interactivellama** [comment]
Yes, I do from time to time. I'm one of the many Midwestern imports into Salesforce. However, my project is an internal open source project. I do hear rumors that the SLDS website will eventually use it for component examples, so no need to check those right now.

**Comment by jessebeach** [comment]
Neat. We'll make sure the update rules work on these examples. The goal is fewer warnings, but for those that do trigger, they'll be actionable and meaningful.

**Comment by jessebeach** [comment]
This improvement has been rolled into the master branch: #194

**Comment by yantakus** [comment]
I've upgraded to v4.0.0 but adding `role="button"` to an element with `onClick` handler doesn't suppress the warning. Is it included in the latest release?

**Comment by jessebeach** [comment]
It'll be in the imminent 5.0 release!

**Comment by yantakus** [comment]
Why is it stated in the docs then: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-static-element-interactions.md ? It is very confusing...

**Comment by ljharb** [comment]
@yantakus the docs on master don't reflect the last published version. You want the docs on the latest git tag: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/4.0.0/docs/rules/no-static-element-interactions.md

**Comment by nnjpp** [comment]
Thanks! I usually work on projects where `master` is stable and delivered so that confused me a lot too.

**Comment by ljharb** [comment]
That's the case for apps, but is almost never the case for published modules/packages/gems/artifacts.

**Comment by Donaldini** [comment]
@ljharb Is there a more specific timeframe than _imminent_ on the 5.0 release?
I'm asking because I don't see any milestones in this project or reference to upcoming releases.

**Comment by evcohen** [comment]
@Donaldini today!
